src/router/open-plc-utils/ether/efsu.1
.TH efsu 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
efsu - Ethernet Frame Send Utility

.SH SYNOPSIS
.BR efsu
.RI [ options ]
.IR file
.RI [ file ]
[...]

.PP
.BR efsu
.RI [ options ]
.RI [< stdin ]

.SH DESCRIPTION

.PP
Sends custom Ethernet frames over a specific host Ethernet interface.

.PP
A frame is specified as a text stream of hexadecimal octets read from one or more frame description files or from stdin.
Each octet is converted to a binary byte, buffered and transmitted over the network on the specified host interface.
The program assumes that frames are valid and will transmit whatever you specify, right or wrong.
The program only checks minimum and maximum frame length.

.PP
Frame descriptions are stored as plain text files.
Multiple frames may be specified in one file by separating them with a semicolon.
All descriptions in a file are read and sent immediately but multiple files may be specified on the command line and options allow pauses between files, command line repetition and pauses between command line iterations.

.PP
A frame descriptions is a continuous stream of hexadecimal digits.
Intervening text, punctuation, white space and comments are discarded.
Hex characters inside script-style, C-style or C++-style comments are ignored.
Script-style comments include all characters between a hash ('\fB#\fR') and the next newline or EOF.
C-style and C++-style comments follow normal language rules.
Comments may be used to document frame descriptions or hide frame segments.
In the end, there must be an even number of hex digits that represent 60 to 1518 octets;

.SH OPTIONS

.TP
-\fBd \fIaddress\fR
Replace the destination Ethernet address of each transmitted frame with the option argument.
The destination address consists of 12 hex digits in upper, lower or mixed case.
Octets may be seperated with colons for clarity but colons are not required.

.TP
-\fBe \fIethertype\fR
Echo received frames having \fIethertype\fR.
This effectively filters received frames but does not filter transmitted frames.
The ethertype is expressed in hexadecimal without the "\fB0x\fR" prefix.
The keyword "\fBhp10\fR" can be used for \fB887B\fR and keyword "\fBhpav\fR" can be used for \fB88E1\fR.

.TP
.RB - h
Replace the source address of each transmitted frame with the Ethernet address of the host interface used to send the frame.
This allows frames to be sent correctly from any host interface.
Use option \fB-i\fR to specify an interface.

.TP
-\fBi \fIinterface\fR
The host interface used to send frames.
The default is "\fBeth1\fR" because most people use "eth0" as their principle network connection; however, if the string "\fBPLC\fR" is defined in the environment then it will over-ride the program default.
Any interface specified on the command line will over-ride the default.

.TP
-\fBl \fIcount\fR
Repeat command line \fIcount\fR times by re-sending all files each time.
The default is \fB1\fR.

.TP
-\fBp \fIseconds\fR
Pause between each file on the command line.
The default is \fB0\fR seconds.

.TP
-\fBt \fItimer\fR
Read timeout in milliseconds.
Values range from 0 through UINT_MAX.
This is the maximum time allowed for a response.
The default is 50 milliseconds.

.TP
.RB - v
Dumps transmitted frames on stdout as they are sent.
If an Ethertype was specified using option \fB-e\fR the received frames having that Ethertype will also be dumped.

.TP
-\fBw \fIseconds\fR
Wait between command line iterations.
The default is \fB0\fR seconds.

.SH ARGUMENTS

.TP
.IR file
The name of a file containing one or more frame specifications.
When more than one file is specified, \fBefsu\fR behaves like the \fBcat\fR utility by reading and sending each file in turn.
The program makes no assumptions based on filename or extension and enforces no filename conventions; however, giving frame description files the *.hex extension is a good convention.
Some toolkit distributions include example \fB.hex\fR files in the source folder.
These files are not needed for proper execution and may be deleted.

.SH NOTES
One way to create a frame description file is to cut-and-paste output from a packet sniffer program and discard the line numbers and ASCII text columns.
You can then break out selected frame segments onto new lines and insert comments.

.SH EXAMPLES
For example, a simple file might look like:

.PP
   #!/usr/local/bin/efsu -ieth0
   # file: abc.hex
   # =====================================
   # this is an example frame description;
   # -------------------------------------
   00 B0 52 00 00 01 
   // 00 B0 52 BE DE 73 
   00 E0 45 DA DA 07
   88 E1 
   /* content */
   00 00 00 00 00 00 00 00
   00 00 00 00 00 00 00 00
   00000000 0000 00 /* control byte */ 00
   00 00 00 00 00 00 00 00
   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
   00 00 00 00 00 00 00 00

.PP
The bang path permits direct execution of the file if 1) \fBefsu\fR is installed in the bang path directory, 2) the description file has execute permissions and 3) the user has execute privileges for that file.
The script-style, C++-style and C-style comments are ignored along with any white space.
Only the hexadecimal octets are read and sent.
Since the number of octets exceeds 60 and the number of digits is even, \fBefsu\fR will report no errors and send the packet.
There is no guarantee that this packet is valid.

.PP
Assuming the above file is named \fBabc.hex\fR, we can transmit the file over the network as an Ethernet frame with the next command.
A hexadecimal dump of the transmitted frame appears on the console because option \fB-v\fR is present.

.PP
   # efsu abc.hex -v
   00000000 00 B0 52 00 00 01 00 E0 45 DA DA 07 88 E1 00 00 ..R.....E.......
   00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000040 00 00 00 00 00 00                               ......

.PP
Since every host interface has a different hardware address, one can replace the source address of transmitted frames with the actual host interface address using option \fB-h\fR.
Observe that the original source address has been replaced with that of the local host interface used to transmit the frame.

.PP
   # efsu abc.hex -vh
   00000000 00 B0 52 00 00 01 \fB00 0F EA 10 D5 1C\fR 88 E1 00 00 ..R.............
   00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000040 00 00 00 00 00 00                               ......

.PP
You can also direct the frame to another device using option \fB-d\fR.
Observe that the original destination address has been replaced with the one specified on the command line.

.PP
   # efsu abc.hex -vhd 00b052beef01
   00000000 \fB00 B0 52 BE EF 01\fR 00 0F EA 10 D5 1C 88 E1 00 00 ..R.............
   00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
   00000040 00 00 00 00 00 00                               ......

.SH SEE ALSO
.BR amp ( 1 ),
.BR hpav ( 1 )

.SH CREDITS
 Charles Maier <cmaier@qca.qualcomm.com>



src/router/open-plc-utils/ether/efsu.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or 
 *   without modification, are permitted (subject to the limitations 
 *   in the disclaimer below) provided that the following conditions 
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright 
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above 
 *     copyright notice, this list of conditions and the following 
 *     disclaimer in the documentation and/or other materials 
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of 
 *     its contributors may be used to endorse or promote products 
 *     derived from this software without specific prior written 
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE 
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE 
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR 
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER 
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT 
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) 
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE 
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.  
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   efsu.c - Ethernet Frame Send Utility;
 *
 *   convert hexadecimal text files to ethernet frames and transmit
 *   them over the network; basically, it is a 'send your own frame'
 *   utility;
 *
 *   the program works like cat, sending file after file to a given
 *   interface; as each file is read, all hexadecimal octets in the
 *   file are converted to bytes and buffered; a semicolon causes a
 *   buffer transmit as does the end of file; script-style comments
 *   starting with hash (#) and c-language-style comments starting
 *   with slash-slash or slash-asterisk are consumed and discard as
 *   the file is read; the errors that can occur are non-hex digits
 *   and odd number of hex digits;
 *
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <stdlib.h>
#include <string.h>
#include <limits.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/error.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/flags.h"
#include "../ether/channel.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/error.c"
#include "../tools/hexencode.c"
#include "../tools/hexload.c"
#include "../tools/hexdump.c"
#include "../tools/todigit.c"
#include "../tools/uintspec.c"
#include "../tools/basespec.c"
#include "../tools/synonym.c"
#endif

#ifndef MAKEFILE
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#include "../ether/channel.c"
#endif

/*====================================================================*
 *   program contants;
 *--------------------------------------------------------------------*/

#define EFSU_INTERFACE "PLC"
#define EFSU_ETHERTYPE 0x88E1
#define EFSU_PAUSE 0
#define EFSU_DELAY 0
#define EFSU_LOOP 1

#ifndef ETHER_CRC_LEN
#define ETHER_CRC_LEN 4
#endif

/*====================================================================*
 *   program variables;
 *--------------------------------------------------------------------*/

static const struct _term_ protocols [] =

{
	{
		"hp10",
		"887B"
	},
	{
		"hpav",
		"88E1"
	}
};


/*====================================================================*
 *
 *   void function (struct channel * channel, void * memory, ssize_t extent);
 *
 *   read Ethernet frame descriptions from stdin and transmit them
 *   as raw ethernet frames; wait for a response if CHANNEL_LISTEN
 *   flagword bit is set;
 *
 *
 *--------------------------------------------------------------------*/

static void function (struct channel * channel, void * memory, ssize_t extent)

{
	struct ether_header * frame = (struct ether_header *)(memory);
	unsigned length;
	while ((length = (unsigned)(hexload (memory, extent, stdin))) > 0)
	{
		if (length < (ETHER_MIN_LEN - ETHER_CRC_LEN))
		{
			error (1, ENOTSUP, "Frame size of %d is less than %d bytes", length, (ETHER_MIN_LEN - ETHER_CRC_LEN));
		}
		if (length > (ETHER_MAX_LEN - ETHER_CRC_LEN))
		{
			error (1, ENOTSUP, "Frame size of %d is more than %d bytes", length, (ETHER_MAX_LEN - ETHER_CRC_LEN));
		}
		if (_anyset (channel->flags, CHANNEL_UPDATE_TARGET))
		{
			memcpy (frame->ether_dhost, channel->peer, sizeof (frame->ether_dhost));
		}
		if (_anyset (channel->flags, CHANNEL_UPDATE_SOURCE))
		{
			memcpy (frame->ether_shost, channel->host, sizeof (frame->ether_shost));
		}
		sendpacket (channel, memory, length);
		if (_anyset (channel->flags, CHANNEL_LISTEN))
		{
			while (readpacket (channel, memory, extent) > 0);
		}
	}
	return;
}


/*====================================================================*
 *
 *   void iterate (int argc, char const * argv [], void * memory, ssize_t extent, unsigned pause);
 *
 *
 *
 *--------------------------------------------------------------------*/

static void iterate (int argc, char const * argv [], struct channel * channel, unsigned pause)

{
	byte buffer [ETHER_MAX_LEN];
	if (!argc)
	{
		function (channel, buffer, sizeof (buffer));
	}
	while ((argc) && (* argv))
	{
		if (!freopen (* argv, "rb", stdin))
		{
			error (1, errno, "Can't open %s", * argv);
		}
		function (channel, buffer, sizeof (buffer));
		argc--;
		argv++;
		if ((argc) && (* argv))
		{
			sleep (pause);
		}
	}
	return;
}


/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"d:e:hi:l:p:t:vw:",
		PUTOPTV_S_FUNNEL,
		"Ethernet Frame Send Utility",
		"d x\treplace destination address with (x)",
		"e x\techo return frames having ethertype (x) [" LITERAL (EFSU_ETHERTYPE) "]",
		"h\treplace source address with host address",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"l n\trepeat file sequence (n) times [" LITERAL (EFSU_LOOP) "]",
		"p n\twait (n) seconds between files [" LITERAL (EFSU_PAUSE) "]",
		"t n\tread timeout is (n) milliseconds [" LITERAL (CHANNEL_TIMEOUT) "]",
		"v\tverbose messages",
		"w n\twait (n) seconds between loops [" LITERAL (EFSU_DELAY) "]",
		(char const *) (0)
	};
	unsigned pause = EFSU_PAUSE;
	unsigned delay = EFSU_DELAY;
	unsigned loop = EFSU_LOOP;
	signed c;
	channel.type = EFSU_ETHERTYPE;
	if (getenv (EFSU_INTERFACE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (EFSU_INTERFACE));

#else

		channel.ifname = strdup (getenv (EFSU_INTERFACE));

#endif

	}
	optind = 1;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'd':
			_setbits (channel.flags, CHANNEL_UPDATE_TARGET);
			if (!hexencode (channel.peer, sizeof (channel.peer), optarg))
			{
				error (1, errno, "%s", optarg);
			}
			break;
		case 'e':
			_setbits (channel.flags, CHANNEL_LISTEN);
			channel.type = (uint16_t)(basespec (synonym (optarg, protocols, SIZEOF (protocols)), 16, sizeof (channel.type)));
			break;
		case 'h':
			_setbits (channel.flags, CHANNEL_UPDATE_SOURCE);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'l':
			loop = (unsigned)(uintspec (optarg, 0, UINT_MAX));
			break;
		case 'p':
			pause = (unsigned)(uintspec (optarg, 0, 1200));
			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			break;
		case 't':
			channel.timeout = (signed)(uintspec (optarg, 0, UINT_MAX));
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			break;
		case 'w':
			delay = (unsigned)(uintspec (optarg, 0, 1200));
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	openchannel (&channel);
	while (loop--)
	{
		iterate (argc, argv, &channel, pause);
		if (loop)
		{
			sleep (delay);
		}
	}
	closechannel (&channel);
	return (0);
}

